wallet/controllers_v3.go

```diff
@@ -0,0 +1,445 @@
+	)
+
+	// get logger from context
+	logger, err := appctx.GetLogger(ctx)
+	if err != nil {
+		// no logger, setup
+		ctx, logger = logging.SetupLogger(ctx)
+	}
+
+	if err := inputs.DecodeAndValidateReader(r.Context(), bcr, r.Body); err != nil {
+		return bcr.HandleErrors(err)
+	}
+
+	var (
+		db Datastore
+		ok bool
+	)
+
+	// get datastore from context
+	if db, ok = ctx.Value(appctx.DatastoreCTXKey).(Datastore); !ok {
+		logger.Error().Msg("unable to get datastore from context")
+		return handlers.WrapError(err, "misconfigured datastore", http.StatusServiceUnavailable)
+	}
+
+	var altCurrency = altcurrency.BAT
+
+	var info = &walletutils.Info{
+		ID:          uuid.NewV4().String(),
+		Provider:    "brave",
+		PublicKey:   publicKey,
+		AltCurrency: &altCurrency,
+	}
+
+	// get wallet from datastore
+	err = db.InsertWallet(info)
+	if err != nil {
+		logger.Error().Err(err).Str("id", info.ID).Msg("unable to create brave wallet")
+		return handlers.WrapError(err, "error writing wallet to storage", http.StatusServiceUnavailable)
+	}
+
+	return handlers.RenderContent(ctx, infoToResponseV3(info), w, http.StatusCreated)
+}
+
+// ClaimUpholdWalletV3 - produces an http handler for the service s which handles claiming of uphold wallets
+func ClaimUpholdWalletV3(s *Service) func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+	return func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		var (
+			ctx = r.Context()
+			id  = new(inputs.ID)
+			cuw = new(ClaimUpholdWalletRequest)
+		)
+		// get logger from context
+		logger, err := appctx.GetLogger(ctx)
+		if err != nil {
+			// no logger, setup
+			ctx, logger = logging.SetupLogger(ctx)
+		}
+
+		// get payment id
+		if err := inputs.DecodeAndValidateString(context.Background(), id, chi.URLParam(r, "paymentID")); err != nil {
+			logger.Warn().Str("paymentID", err.Error()).Msg("failed to decode and validate paymentID from url")
+			return handlers.ValidationError(
+				"error validating paymentID url parameter",
+				map[string]interface{}{
+					"paymentID": err.Error(),
+				},
+			)
+		}
+
+		// read post body
+		if err := inputs.DecodeAndValidateReader(r.Context(), cuw, r.Body); err != nil {
+			return cuw.HandleErrors(err)
+		}
+
+		// get the wallet
+		wallet, err := s.GetWallet(id.UUID())
+		if err != nil {
+			if strings.Contains(err.Error(), "looking up wallet") {
+				return handlers.WrapError(err, "unable to find wallet", http.StatusNotFound)
+			}
+			return handlers.WrapError(err, "unable to get or create wallets", http.StatusServiceUnavailable)
+		}
+
+		var aa uuid.UUID
+
+		if cuw.AnonymousAddress != "" {
+			aa, err = uuid.FromString(cuw.AnonymousAddress)
+			if err != nil {
+				return handlers.ValidationError(
+					"error validating anonymousAddress",
+					map[string]interface{}{
+						"anonymousAddress": err.Error(),
+					},
+				)
+			}
+		}
+
+		var amount decimal.Decimal
+
+		if wallet.ProviderID == "" && cuw.SignedLinkingRequest != "" {
```

Comment:
I think the naming of this function is still throwing me off. We're not assured to be claiming an "uphold wallet" as in a wallet with provider = uphold. Instead we're claiming a wallet (whomever the provider is ) using an uphold user account right? Thus this branch is handling the provider = "brave" case?

Reply:
You are saying that `ClaimUpholdWalletV3` is not really accurate?  How about `LinkDepositAccountV3` instead, as that is much more accurate of what this is doing I think?

Reply:
Here is a PR that changes the name (base branch is add-wallet-endpoints).  Does this help clarify the naming issue?

https://github.com/brave-intl/bat-go/pull/446